File: Examples/Objective-C/Remote Control/Remote Control/MainViewController.m
//
//  ViewController.m
//  Remote Control
//
//  Created by Calvin on 7/7/17.
//  Copyright © 2017 Bridgefy Inc. All rights reserved.
//

#import "MainViewController.h"
#import "AdminViewTabBarController.h"
#import "ImagePickerViewController.h"
#import "ColorPickerViewController.h"
#import "InputTextViewController.h"
#import "FlashlightViewController.h"

#import <BFTransmitter/BFTransmitter.h>
#import <AVFoundation/AVFoundation.h>

#define kLastId @"lastId"
#define kCommandKey @"command"
#define kIdKey @"id"
#define kImageKey @"image"
#define kColorKey @"color"
#define kTextKey @"text"

typedef NS_ENUM(NSUInteger, Command) {
    CommandImage = 1,
    CommandColor,
    CommandFlashlight,
    CommandText
};

@interface MainViewController () <BFTransmitterDelegate> {
    NSArray *images;
}

@property (nonatomic, retain) BFTransmitter *transmitter;

@property (weak, nonatomic) IBOutlet UIImageView *imageView;
@property (weak, nonatomic) IBOutlet UILabel *messageLabel;

@end

@implementation MainViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    images = @[@"ad", @"sports", @"map", @"concert"];
    
    self.navigationItem.backBarButtonItem = [[UIBarButtonItem alloc] initWithTitle:@""
                                                                             style:UIBarButtonItemStylePlain
                                                                            target:nil
                                                                            action:nil];
    
    //Transmitter initialization
    [BFTransmitter setLogLevel:BFLogLevelError];
    self.transmitter = [[BFTransmitter alloc] initWithApiKey:@"YOUR API KEY"];
    self.transmitter.delegate = self;
    self.transmitter.backgroundModeEnabled = YES;
    [self.transmitter start];
    
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (IBAction)longPressDetected:(UILongPressGestureRecognizer *)gesture {
    if (gesture.state == UIGestureRecognizerStateBegan) {
        [self showAdminDialog];
    }
}

- (void)showAdminDialog {
    UIAlertController *alertController = [UIAlertController alertControllerWithTitle:@"Do you want to become an admin?"
                                                                             message:@""
                                                                      preferredStyle:UIAlertControllerStyleAlert];
    
    UIAlertAction *cancelAction = [UIAlertAction actionWithTitle:@"Cancel"
                                                           style:UIAlertActionStyleCancel
                                                         handler:nil];
    
    UIAlertAction *okAction = [UIAlertAction actionWithTitle:@"OK"
                                                       style:UIAlertActionStyleDefault
                                                     handler:^(UIAlertAction * _Nonnull action) {
                                                         [self performSegueWithIdentifier:@"showAdminView"
                                                                                   sender:self];
                                                     }];
    
    [alertController addAction:cancelAction];
    [alertController addAction:okAction];
    
    [self presentViewController:alertController
                       animated:YES
                     completion:nil];
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    if ([segue.identifier isEqualToString:@"showAdminView"]) {
        AdminViewTabBarController *adminVC = (AdminViewTabBarController *)segue.destinationViewController;
        adminVC.mvc = self;
    }
}

- (void)sendObject:(id)object sender:(id)sender {
    NSMutableDictionary *dict = [[NSMutableDictionary alloc] init];
    [dict setObject:@(floor([[NSDate date] timeIntervalSince1970] * 1000)) forKey:kIdKey];
    
    if ([sender isKindOfClass:[ImagePickerViewController class]]) {
        [dict setObject:@(CommandImage) forKey:kCommandKey];
        [dict setObject:object forKey:kImageKey];
    } else if ([sender isKindOfClass:[ColorPickerViewController class]]) {
        [dict setObject:@(CommandColor) forKey:kCommandKey];
        [dict setObject:object forKey:kColorKey];
    } else if ([sender isKindOfClass:[FlashlightViewController class]]) {
        [dict setObject:@(CommandFlashlight) forKey:kCommandKey];
    } else if ([sender isKindOfClass:[InputTextViewController class]]) {
        [dict setObject:@(CommandText) forKey:kCommandKey];
        [dict setObject:(NSString *)object forKey:kTextKey];
    } else {
        NSLog(@"ERROR: Unknown sender");
        return;
    }
    
    BFSendingOption options = BFSendingOptionMeshTransmission|BFSendingOptionBroadcastReceiver|BFSendingOptionNotEncrypted;
    
    NSError *error;
    [self.transmitter sendDictionary:dict
                              toUser:nil
                             options:options
                               error:&error];
}

#pragma mark - BFTransmitterDelegate

- (void)transmitter:(BFTransmitter *)transmitter meshDidAddPacket:(NSString *)packetID {
    //Packet added to mesh
}

- (void)transmitter:(BFTransmitter *)transmitter didReachDestinationForPacket:( NSString *)packetID {
    //Mesh packet reached destiny (no always invoked)
}

- (void)transmitter:(BFTransmitter *)transmitter meshDidStartProcessForPacket:( NSString *)packetID {
    //A message entered in the mesh process.
}
- (void)transmitter:(BFTransmitter *)transmitter didSendDirectPacket:(NSString *)packetID{
    //A direct message was sent
}

- (void)transmitter:(BFTransmitter *)transmitter didFailForPacket:(NSString *)packetID error:(NSError * _Nullable)error {
    //A direct message transmission failed.
}
- (void)transmitter:(BFTransmitter *)transmitter meshDidDiscardPackets:(NSArray<NSString *> *)packetIDs {
    //A mesh message was discared and won't still be transmitted.
}

- (void)transmitter:(BFTransmitter *)transmitter meshDidRejectPacketBySize:(NSString *)packetID {
    NSLog(@"The packet %@ was rejected from mesh because it exceeded the limit size.", packetID);
}

- (void)transmitter:(BFTransmitter *)transmitter
didReceiveDictionary:(NSDictionary<NSString *, id> * _Nullable) dictionary
           withData:(NSData * _Nullable)data
           fromUser:(NSString *)user
           packetID:(NSString *)packetID
          broadcast:(BOOL)broadcast
               mesh:(BOOL)mesh {
    // A dictionary was received by BFTransmitter.
    
    [self processReceivedDictionary:dictionary];
    
}

- (void)transmitter:(BFTransmitter *)transmitter didDetectConnectionWithUser:(NSString *)user {
    //A connection was detected (no necessarily secure)
}

- (void)transmitter:(BFTransmitter *)transmitter didDetectDisconnectionWithUser:(NSString *)user {
    // A disconnection was detected.
}

- (void)transmitter:(BFTransmitter *)transmitter didFailAtStartWithError:(NSError *)error {
    NSLog(@"An error occurred at start: %@", error.localizedDescription);
}

- (void)transmitter:(BFTransmitter *)transmitter didOccurEvent:(BFEvent)event description:(NSString *)description {
    NSLog(@"Event reported: %@", description);
}

- (BOOL)transmitter:(BFTransmitter *)transmitter shouldConnectSecurelyWithUser:(NSString *)user {
    return NO; //if YES establish connection with encryption capacities.
}

- (void)transmitter:(BFTransmitter *)transmitter didDetectSecureConnectionWithUser:(nonnull NSString *)user {
    // A secure connection was detected,
    // A secure connection has encryption capabilities.
}

#pragma mark - 

- (void)processReceivedDictionary:(NSDictionary *)dict {
    
    double receivedId = [[dict valueForKey:kIdKey] doubleValue];
    
    if (![self updateLastId:receivedId]) {
        // Command is ignored
        return;
    }
    
    Command cmd = (Command)[[dict valueForKey:kCommandKey] intValue];
    
    switch (cmd) {
        case CommandImage:
            [self showImageFromDictionary:dict];
            break;
            
        case CommandColor:
            [self showColorFromDictionary:dict];
            break;
            
        case CommandFlashlight:
            [self turnOnFlashlight:YES];
            break;
            
        case CommandText:
            [self showTextFromDictionary:dict];
            break;
    }
}

- (void)showImageFromDictionary:(NSDictionary *)dict {
    int imageIndex = [[dict valueForKey:kImageKey] intValue];
    
    if (imageIndex > images.count) {
        return;
    }
    
    [self resetView];
    self.imageView.image = [UIImage imageNamed:images[imageIndex]];
    self.imageView.hidden = NO;
    
}

- (void)showColorFromDictionary:(NSDictionary *)dict {
    int c = [[dict valueForKey:kColorKey] intValue];
    UIColor *receivedColor = [UIColor colorWithRed:((c >> 16) & 0xFF) / 255.0
                                             green:((c >> 8) & 0xFF) / 255.0
                                              blue:(c & 0xFF) / 255.0
                                             alpha:((c >> 24) & 0xFF) / 255.0];
    
    [self resetView];
    self.view.backgroundColor = receivedColor;
}

- (void)turnOnFlashlight:(BOOL)flag {
    AVCaptureDevice *flashlight = [AVCaptureDevice defaultDeviceWithMediaType:AVMediaTypeVideo];
    
    if (!flashlight) {
        NSLog(@"Can't play with torch");
        return;
    }
    
    if (flashlight.isTorchAvailable &&
        flashlight.hasTorch) {
        NSError *error;
        [flashlight lockForConfiguration:&error];
        
        if (error) {
            NSLog(@"ERROR: %@", error.localizedDescription);
            return;
        }
        
        if (flag) {
            if (flashlight.torchMode == AVCaptureTorchModeOn) {
                // If flashlight is turned on, the command is ignored
                return;
            } else {
                flashlight.torchMode = AVCaptureTorchModeOn;
                [self resetView];
                self.imageView.image = [UIImage imageNamed:@"Flashlight"];
                self.imageView.hidden = NO;
                [self performSelector:@selector(turnOffFlashlight)
                           withObject:nil
                           afterDelay:15.0];
            }
        } else {
            flashlight.torchMode = AVCaptureTorchModeOff;
            [self resetView];
        }
        
        [flashlight unlockForConfiguration];
        
    }
}

- (void)turnOffFlashlight {
    [self turnOnFlashlight:NO];
}

- (void)showTextFromDictionary:(NSDictionary *)dict {
    NSString *text = [dict objectForKey:kTextKey];
    self.messageLabel.text = text;
    self.messageLabel.font = [UIFont boldSystemFontOfSize:20.0];
    
    [self resetView];
    self.messageLabel.hidden = NO;
}

- (void)resetView {
    self.imageView.hidden = YES;
    self.messageLabel.hidden = YES;
    self.view.backgroundColor = [UIColor whiteColor];
}

- (BOOL)updateLastId:(double)receivedId {
    NSUserDefaults *userDefaults = [NSUserDefaults standardUserDefaults];
    
    double savedId = [[userDefaults valueForKey:kLastId] doubleValue];
    
    if (receivedId > savedId) {
        [userDefaults setObject:@(receivedId) forKey:kLastId];
        [userDefaults synchronize];
        return YES;
    }
    
    return NO;
    
}

@end

File: Examples/Swift/Remote Control/Remote Control/MainViewController.swift
//
//  MainViewController.swift
//  Remote Control
//
//  Created by Calvin on 7/11/17.
//  Copyright © 2017 Bridgefy Inc. All rights reserved.
//

import UIKit
import BFTransmitter
import AVFoundation

let kLastID = "lastId"
let kCommandKey = "command"
let kIdKey = "id"
let kImageKey = "image"
let kColorKey = "color"
let kTextKey = "text"

enum Comand : Int {
    case image = 1
    case color
    case flashlight
    case text
}

class MainViewController: UIViewController, BFTransmitterDelegate {
    
    let images = ["ad", "sports", "map", "concert"]
    var transmitter: BFTransmitter?
    
    @IBOutlet weak var imageView: UIImageView!
    @IBOutlet weak var messageLabel: UILabel!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.navigationItem.backBarButtonItem = UIBarButtonItem(title: "", style: .plain, target: nil, action: nil);
        
        // Transmitter initialization
        BFTransmitter.setLogLevel(.error)
        self.transmitter = BFTransmitter(apiKey: "YOUR API KEY")
        self.transmitter?.delegate = self
        self.transmitter?.isBackgroundModeEnabled = true
        self.transmitter?.start()
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    @IBAction func longPressDetected(_ sender: UILongPressGestureRecognizer) {
        if sender.state == .began {
            self.showAdminDialog()
        }
    }
    
    func showAdminDialog() {
        let alertController = UIAlertController(title: "Do you want to become an admin?",
                                                message: "",
                                                preferredStyle: .alert)
        
        let cancelAction = UIAlertAction(title: "Cancel",
                                         style: .cancel,
                                         handler: nil)
        
        let okAction = UIAlertAction(title: "OK", style: .default) { (action) in
            self.performSegue(withIdentifier: "showAdminView", sender: self)
        }
        
        alertController.addAction(cancelAction)
        alertController.addAction(okAction)
        
        self.present(alertController, animated: true)
    }
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        if segue.identifier == "showAdminView" {
            let adminVC = segue.destination as! AdminViewTabBarController
            adminVC.mvc = self
        }
    }
    
    public func send(object: Any?, sender: Any) {
        var dict: [String: Any] = Dictionary.init()
        
        dict[kIdKey] = floor(Date.init().timeIntervalSince1970 * 1000) as Any
        
        if sender is ImagePickerViewController {
            dict[kCommandKey] = Comand.image.rawValue as Any
            dict[kImageKey] = object
        } else if sender is ColorPickerViewController {
            dict[kCommandKey] = Comand.color.rawValue as Any
            dict[kColorKey] = object
        } else if sender is FlashlightViewController {
            dict[kCommandKey] = Comand.flashlight.rawValue as Any
        } else if sender is InputTextViewController {
            dict[kCommandKey] = Comand.text.rawValue as Any
            dict[kTextKey] = object
        } else {
            print("ERROR: Unknown sender")
            return;
        }
        
        let options: BFSendingOption = [.broadcastReceiver, .meshTransmission]
        
        do {
            try self.transmitter?.send(dict, toUser: nil, options: options)
        } catch let err as NSError {
            print("ERROR: \(err.localizedDescription)")
        }
        
        
    }
    
    // MARK: - BFTransmitterDelegate
    
    public func transmitter(_ transmitter: BFTransmitter, meshDidAddPacket packetID: String) {
        
    }
    
    public func transmitter(_ transmitter: BFTransmitter, didReachDestinationForPacket packetID: String) {
        //Mesh packet reached destiny (no always invoked)
    }
    
    public func transmitter(_ transmitter: BFTransmitter, meshDidStartProcessForPacket packetID: String) {
        //A message entered in the mesh process (was added).
        // Just called when the option BFSendingOptionFullTransmission was used.
    }
    
    public func transmitter(_ transmitter: BFTransmitter, didSendDirectPacket packetID: String) {
        //A direct message was sent
    }
    
    public func transmitter(_ transmitter: BFTransmitter, didFailForPacket packetID: String, error: Error?) {
        //A direct message transmission failed.
    }
    
    public func transmitter(_ transmitter: BFTransmitter, meshDidDiscardPackets packetIDs: [String]) {
        //A mesh message was discared and won't still be transmitted.
        
    }
    
    public func transmitter(_ transmitter: BFTransmitter, meshDidRejectPacketBySize packetID: String) {
        print("The packet \(packetID) was rejected from mesh because it exceeded the limit size.");
    }
    
    public func transmitter(_ transmitter: BFTransmitter,
                            didReceive dictionary: [String : Any]?,
                            with data: Data?,
                            fromUser user: String,
                            packetID: String,
                            broadcast: Bool,
                            mesh: Bool) {
        // A dictionary was received by BFTransmitter.
        
        self.processReceived(dict: dictionary)
        
        
    }
    
    public func transmitter(_ transmitter: BFTransmitter, didDetectConnectionWithUser user: String) {
        //A connection was detected (no necessarily secure)
    }
    
    public func transmitter(_ transmitter: BFTransmitter, didDetectDisconnectionWithUser user: String) {
        // A disconnection was detected.
    }
    
    public func transmitter(_ transmitter: BFTransmitter, didFailAtStartWithError error: Error)
    {
        print("An error occurred at start: \(error.localizedDescription)");
    }
    
    public func transmitter(_ transmitter: BFTransmitter, didOccur event: BFEvent, description: String)
    {
        print("Event reported: \(description)");
    }
    
    public func transmitter(_ transmitter: BFTransmitter, shouldConnectSecurelyWithUser user: String) -> Bool {
        return false//if True, it will establish connection with encryption capacities.
        // Not necessary for this case.
    }
    
    public func transmitter(_ transmitter: BFTransmitter, didDetectSecureConnectionWithUser user: String) {
        // A secure connection was detected,
    }
    
    // MARK: -
    
    func processReceived(dict: [String: Any]?) {
        
        guard let dict = dict else {
            return
        }
        
        let receivedID = dict[kIdKey] as! Double
        
        if !self.update(lastId: receivedID) {
            // Command is ignored
            return
        }
        
        let cmd: Comand = Comand(rawValue: dict[kCommandKey] as! Int)!
        
        switch cmd {
        case .image:
            self.showImage(from: dict)
        case .color:
            self.showColor(from: dict)
        case .flashlight:
            self.turnOnFlashlight(flag: true)
        case .text:
            self.showText(from: dict)
        }
        
    }
    
    func showImage(from dictionary: [String: Any]) {
        let imageIndex = dictionary[kImageKey] as! Int
        
        if imageIndex > self.images.count {
            return
        }
        
        self.resetView()
        self.imageView.image = UIImage(named: self.images[imageIndex])
        self.imageView.isHidden = false
    }
    
    func showColor(from dictionary: [String: Any]) {
        let c = dictionary[kColorKey] as! Int
        let receivedColor = UIColor(colorLiteralRed: Float((c >> 16) & 0xFF) / 255.0,
                                    green: Float((c >> 8) & 0xFF) / 255.0,
                                    blue: Float(c & 0xFF) / 255.0,
                                    alpha: Float((c >> 24) & 0xFF) / 255.0)
        
        self.resetView()
        self.view.backgroundColor = receivedColor
    }
    
    func turnOnFlashlight(flag: Bool) {
        
        guard let flashlight = AVCaptureDevice.defaultDevice(withMediaType: AVMediaTypeVideo) else {
            print("Can't play with torch")
            return
        }
        
        if flashlight.isTorchAvailable && flashlight.hasTorch {
            
            do {
                try flashlight.lockForConfiguration()
            } catch let err as NSError {
                print("ERROR: \(err.localizedDescription)")
                return
            }
            
            if flag {
                
                if flashlight.torchMode == .on {
                    // If flashlight is turned on, the command is ignored
                    return
                } else {
                    flashlight.torchMode = .on
                    self.resetView()
                    self.imageView.image = #imageLiteral(resourceName: "Flashlight")
                    self.imageView.isHidden = false
                    self.perform(#selector(turnOffFlashlight),
                                 with: nil,
                                 afterDelay: 15.0)
                }
                
            } else {
                flashlight.torchMode = .off
                self.resetView()
            }
            
            flashlight.unlockForConfiguration()
        }
    }
    
    func turnOffFlashlight() {
        self.turnOnFlashlight(flag: false)
    }
    
    func showText(from dictionary: [String: Any]) {
        let text = dictionary[kTextKey] as! String
        self.messageLabel.text = text
        self.messageLabel.font = UIFont.boldSystemFont(ofSize: 20.0)
        
        self.resetView()
        self.messageLabel.isHidden = false
    }
    
    func resetView() {
        self.imageView.isHidden = true
        self.messageLabel.isHidden = true
        self.view.backgroundColor = #colorLiteral(red: 1.0, green: 1.0, blue: 1.0, alpha: 1.0)
    }
    
    func update(lastId receivedId: Double) -> Bool {
        let userDefaults = UserDefaults.standard
        
        let savedId = userDefaults.double(forKey: kLastID)
        
        if receivedId > savedId {
            userDefaults.set(receivedId, forKey: kLastID)
            userDefaults.synchronize()
            return true
        }
        
        return false
    }

}
